bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/SortedLedgerStorage.java

```diff
@@ -209,16 +202,12 @@ public void onSizeLimitReached(final Checkpoint cp) throws IOException {
             public void run() {
                 try {
                     LOG.info("Started flushing mem table.");
-                    long logIdBeforeFlush = entryLogger.getCurrentLogId();
                     memTable.flush(SortedLedgerStorage.this);
-                    long logIdAfterFlush = entryLogger.getCurrentLogId();
                     // in any case that an entry log reaches the limit, we roll the log and start checkpointing.
                     // if a memory table is flushed spanning over two entry log files, we also roll log. this is
                     // for performance consideration: since we don't wanna checkpoint a new log file that ledger
                     // storage is writing to.
-                    if (entryLogger.reachEntryLogLimit(0) || logIdAfterFlush != logIdBeforeFlush) {
-                        LOG.info("Rolling entry logger since it reached size limitation");
-                        entryLogger.rollLog();
+                    if (entryLogger.rollLogsIfEntryLogLimitReached()) {
```

Comment:
i don't think it is significant to have any perf impact. First of all this will be called once EntryMemTable reaches size limit, and in this method I'm getting a collection of active entrylogs (in this case just one) and checks its size. For entrylogperledger this collection is required but for single entrylog it might not be required, but I want common implementation and it is called just once per EntryMemTable size limit, so it is not a thing to be concerned about.

Reply:
@reddycharan well, the existence of the code had its reason. performance is one, I think the biggest problem is "checkpoint" logic can potentially never be triggered.

- during memtable flushes, rollLog is disabled for appending memtable entries.
- however background compaction can also advance entrylog, since it is appending entries into same entrylog.
- if a memtable flush spanning over two entry log files, `rollLogsIfEntryLogLimitReached` will most likely be false, so no checkpoint will happen.
- in the next memtable flush, if it is spanning over two entry log files again, then `rollLogsIfEntryLogLimitReached` still be false, no checkpoint will happen.

This was actually a behavior we observed at Twitter 3 years ago. I added this logic to prevent this happen. see twitter/bookkeeper@a7a8c88eaa2659b5d6c4e29f4eb9e2d184c7cfe0

Reply:
brought back rollLogs

Reply:
@sijie if ```rollLogsIfEntryLogLimitReached``` is false, it will fall through to the else if statement, which covers the flush over two logs.

This is weird though, because it is really tied to a single log, but it acts like it's working with many.

Reply:
@ivankelly that's what I commented before. I don't think the methods in EntryLogManager provide the right abstraction. That's what I also dislike your approach on asking moving the class out at this PR. At this PR, we need to figure out what are the right methods to put in an EntryLogManager before any kind of code movement. You comments about inner classes complicate things.

Reply:
> @ivankelly that's what I commented before. I don't think the methods in EntryLogManager provide the right abstraction. 

I agree about the methods providing the right abstraction. I was under the impression that #getPreviousAllocatedEntryLogId() would always return an incremented value if the current log has changed, but having read the code, this is not the case.

It seems to me that SortedLedgerStorage is reaching very far into the internals of the entrylogger to do what it wants to do. Since this is the case, we shouldn't try to shoehorn this into the abstraction. SortedLedgerStorage assumes that it uses a single log, so it should only ever be constructed with a SingleLogEntryLogmanager, and SingleLogEntryLogManager should expose the stuff that SortedLedgerStorage has required until now.